ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/VolDB.hs

```diff
@@ -0,0 +1,340 @@
+import           Ouroboros.Consensus.Util (lastMaybe)
+
+import           Ouroboros.Consensus.Storage.ChainDB.Impl.VolDB
+                     (PointChainDiff (PointChainDiff), fromChainHash,
+                     isReachable)
+import qualified Ouroboros.Consensus.Storage.ChainDB.Impl.VolDB as VolDB
+import qualified Ouroboros.Consensus.Storage.VolatileDB as VolDB
+
+import           Test.Tasty
+import           Test.Tasty.QuickCheck
+
+import           Test.Ouroboros.Storage.TestBlock
+import           Test.Util.Orphans.Slotting.Arbitrary ()
+
+{-------------------------------------------------------------------------------
+  Top-level tests
+-------------------------------------------------------------------------------}
+
+tests :: TestTree
+tests = testGroup "VolDB" [
+      testProperty "isReachable" prop_isReachable
+    ]
+
+{-------------------------------------------------------------------------------
+  Test properties
+-------------------------------------------------------------------------------}
+
+prop_isReachable :: Pretty (ReachableSetup TestBlock) -> Property
+prop_isReachable (Pretty ReachableSetup { currentChain, forkTip, fork, blockInfo }) =
+        Pretty (isReachable (flip Map.lookup blockInfo) currentChain forkTip)
+    === Pretty fork
+
+{-------------------------------------------------------------------------------
+  Test setup
+-------------------------------------------------------------------------------}
+
+data ReachableSetup blk = ReachableSetup {
+      currentChain :: AnchoredFragment (Header blk)
+    , forkTip      :: RealPoint blk
+    , fork         :: Maybe (PointChainDiff blk)
+    , blockInfo    :: Map (HeaderHash blk) (VolDB.BlockInfo (HeaderHash blk))
+    }
+
+deriving instance (HasHeader blk, Show (Header blk)) => Show (ReachableSetup blk)
+
+-- | Newtype that uses 'ppShow' for its 'Show' instance.
+newtype Pretty a = Pretty a
+  deriving (Eq, Arbitrary)
+
+instance Show a => Show (Pretty a) where
+  show (Pretty a) = ppShow a
+
+{-------------------------------------------------------------------------------
+  Generators
+-------------------------------------------------------------------------------}
+
+-- | All blocks on the current chain use this value as for their body. This
+-- allows us to statically convert a header from the current chain to a block.
+currentChainBody :: TestBody
+currentChainBody = TestBody 0 True
+
+genFirstBlock ::
+     Gen TestBody
+  -> Gen TestBlock
+genFirstBlock genBody = oneof
+    [ firstBlock 0            <$> genBody
+    , firstEBB   (const True) <$> genBody
+    ]
+
+genSuccessor ::
+     TestBlock
+  -> Gen TestBody
+  -> Gen TestBlock
+genSuccessor prevBlk genBody = frequency
+    [ ( if prevIsEBB then 0 else 1
+      , mkNextEBB (const True) prevBlk
+          <$> chooseSlot (prevSlot + 1) (prevSlot + 2)
+          -- We don't care about EpochNo
+          <*> return 0
+          <*> genBody
+      )
+    , (3
+      , mkNextBlock prevBlk
```

Comment:
Change `mkNextEBB` and `mkNextBlock` to take `HeaderFields` and `ChainLength`, so that we don't have to do all the faking.